modal/volume.py

```diff
@@ -266,61 +275,164 @@ async def remove_file(self, path: Union[str, bytes], recursive: bool = False) ->
         await retry_transient_errors(self._client.stub.VolumeRemoveFile, req)
 
     @live_method
-    async def _add_local_file(
-        self, local_path: Union[Path, str], remote_path: Optional[Union[str, PurePosixPath, None]] = None
-    ):
-        mount_file = await self._upload_file(local_path, remote_path)
-        request = api_pb2.VolumePutFilesRequest(volume_id=self.object_id, files=[mount_file])
-        await retry_transient_errors(self._client.stub.VolumePutFiles, request, base_delay=1)
+    async def put(self) -> "_VolumeUploadContextManager":
```

Comment:
How about `start_batch()` or `begin()`?

When you said `put()`, I thought it was doing a mutation; a context manager wasn't what I expected!

Reply:
> How about start_upload()?

SGTM.

> Also, with a method name like this, we might be able to skip the _local_* parts of the method names.
> And add_file could then reasonably take either a file-object or a file path.

Only reason for calling them `add_local_file` and `add_local_dir` was consistency with existing NFS APIs:
 * https://github.com/modal-labs/modal-client/blob/06c36d19661c1a91f444d2df523759eac6cc9272/modal/network_file_system.py#L212
 * https://github.com/modal-labs/modal-client/blob/06c36d19661c1a91f444d2df523759eac6cc9272/modal/network_file_system.py#L225

Similarly, `add_fileobj` was named to try to be somewhat consistent with:
 * https://github.com/modal-labs/modal-client/blob/06c36d19661c1a91f444d2df523759eac6cc9272/modal/volume.py#L214

There's also other prior art here (which doesn't accept a file path or file-object):
* https://github.com/modal-labs/modal-client/blob/06c36d19661c1a91f444d2df523759eac6cc9272/modal/network_file_system.py#L147

Reply:
Thanks for mentioning this! Agree there's a tradeoff between consistency with the old API and designing a better new API. I just wonder if we should weigh towards the latter because:

- The old API is already inconsistent with itself in places: https://linear.app/modal-labs/issue/MOD-1731/volume-and-nfs-methods-are-inconsistent
- NFS objects have problems with cross-region peering / latency and we're trying to reduce usage
- This API is behind a separate batched interface which is different from the current one.

Reply:
Favouring "nice API" over consistency since we're trying to move away from NFS is a good point :+1:

Reply:
Would it make sense for the batch interface to expose the same methods (and signatures) as the volume interface? e.g. 

```
vol.put_file(local, remote)
```

and

```
with vol.batch_transfer() as batch:
    batch.put_file(local_a, remote_a)
    batch.put_file(local_b, remote_b)
```

BTW, +1 on having `local` accept either a pathlike or a fileobj wherever possible to align with other Python data  APIs.

Reply:
@mwaskom We currently have a technical limitation that makes sequential, non-batched uploads inefficient. Until that is rectified (not a quick fix), I want to avoid non-batched counterparts for uploads.

Even if we fix that issue, my weak personal preference would still be to only allow uploading via a batched interface. That keeps the API small and will typically perform better (can do uploads etc concurrently) - there's some value value in only having a single way to upload files. Having to write one line extra code also doesn't seem like a huge downside.